src/util/std_expr.h

```diff
@@ -4283,47 +4283,81 @@ class infinity_exprt : public nullary_exprt
 };
 
 /// \brief A let expression
-class let_exprt : public ternary_exprt
+class let_exprt : public binary_exprt
 {
 public:
   let_exprt(symbol_exprt symbol, exprt value, const exprt &where)
-    : ternary_exprt(
+    : binary_exprt(
+        multi_ary_exprt(
+          ID_tuple,
+          {std::move(symbol), std::move(value)},
+          typet(ID_tuple)),
+        ID_let,
+        where,
+        where.type())
+  {
+  }
+
+  using symbol_value_sequencet = std::vector<exprt>;
+
+  let_exprt(symbol_value_sequencet symbol_value_sequence, const exprt &where)
+    : binary_exprt(
+        multi_ary_exprt(
+          ID_tuple,
+          std::move(symbol_value_sequence),
+          typet(ID_tuple)),
         ID_let,
-        std::move(symbol),
-        std::move(value),
         where,
         where.type())
   {
   }
 
+  symbol_value_sequencet &symbol_value_sequence()
```

Comment:
So someone retrieving this needs to know it's an alternating sequence of symbol, value, symbol, value? Seems better to either have a sequence of 2-tuples called bindings, or else provide accessors for nth_symbol, nth_value etc. Either of these can also enforce that the symbols are symbol_exprt.

Reply:
I did consider both options.
Problems of the "accessor" option are that a) you can't easily iterate, and b) it kills the USE_LIST option for performance reasons (but we may want to kill that anyway).
The two-tuple option would be somewhat more memory-consuming, but maybe that's not a genuine issue.

Reply:
Now with the "sequence of 2-tuples" option.